### PR TITLE
[pipelining] add type checking to _backward functions

### PR DESCRIPTION
**Pull request by H-Huang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #140418
* __->__ #140019

fix https://github.com/pytorch/pytorch/issues/139405

cc @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/distributed/pipelining/_backward.py [hunk 1]
```diff
@@ -189,35 +189,30 @@ def hook(grad_inputs):
             handle = intermediate.register_prehook(get_hook(param_group, i))
             handles.append(handle)
 
-    # Stage 0 inputs do not require grads? Should we skip in that case?
-    if all(tensor.requires_grad for tensor in input_values):
```
Comment (wconstab) [inline torch/distributed/pipelining/_backward.py:193]: how come this if condition can be removed now?

File: torch/distributed/pipelining/_backward.py [hunk 2]
```diff
@@ -271,15 +266,15 @@ def stage_backward_weight(
             else:
                 weight.grad += dw
     # return grads in the original order weights were provided in
-    return weight_grads
+    return tuple(weight_grads)
```
Comment (wconstab) [inline torch/distributed/pipelining/_backward.py:269]: these changes look pretty innocuous to me, but can you convince me that this change doesn't add any restriction or limitation to the user code?

**Review by wconstab — APPROVED** [review]
lgtm, thanks!

File: torch/distributed/pipelining/_backward.py (hunk 1 above)
Comment (H-Huang) [reply torch/distributed/pipelining/_backward.py:193]: we no longer call `stage_backward_input` for the first stage anymore

**Pull request closed without merging**